{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparison of position alignment:\n",
      "\n",
      "Number of entries in bwa: 2488\n",
      "Number of entries in custom: 2488\n",
      "% of reads from bwa existing in custom: 1.0\n",
      "% of reads with correct mapping: 1.0\n",
      "Number of reads with wrong number of appearence: 0\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "from operator import itemgetter\n",
    "\n",
    "def parse_result_file_sam(filepath):\n",
    "    reads_info = dict()\n",
    "    with open(filepath) as f:\n",
    "        f.readline()\n",
    "        f.readline()\n",
    "        line = f.readline()\n",
    "        while line:\n",
    "            tokens = line.split(\"\\t\")\n",
    "            reads_info_entry = dict()\n",
    "            reads_info_entry[\"flags\"] = int(tokens[1]) & 0x10\n",
    "            reads_info_entry[\"fasta_seq\"] = tokens[2]\n",
    "            reads_info_entry[\"start_ref\"] = int(tokens[3])\n",
    "            reads_info_entry[\"map_qual\"] = tokens[4]\n",
    "            reads_info_entry[\"cigar\"] = tokens[5]\n",
    "            reads_info_entry[\"length\"] = int(len(tokens[9]))\n",
    "            if tokens[0] in reads_info:\n",
    "                reads_info[tokens[0]].append(reads_info_entry)\n",
    "            else:\n",
    "                reads_info[tokens[0]] = [reads_info_entry]\n",
    "            line = f.readline()\n",
    "\n",
    "    return reads_info\n",
    "\n",
    "\n",
    "def parse_result_file_custom(filepath):\n",
    "    reads_info = dict()\n",
    "    with open(filepath) as f:\n",
    "        line = f.readline()\n",
    "        while line:\n",
    "            tokens = line.split(\"\\t\")\n",
    "            reads_info_entry = dict()\n",
    "            reads_info_entry[\"flags\"] = int(tokens[1])\n",
    "            reads_info_entry[\"fasta_seq\"] = tokens[2]\n",
    "            reads_info_entry[\"start_ref\"] = int(tokens[3])\n",
    "            reads_info_entry[\"map_qual\"] = tokens[4]\n",
    "            reads_info_entry[\"cigar\"] = tokens[5]\n",
    "            reads_info_entry[\"length\"] = int(len(tokens[6]))\n",
    "            if tokens[0] in reads_info:\n",
    "                reads_info[tokens[0]].append(reads_info_entry)\n",
    "            else:\n",
    "                reads_info[tokens[0]] = [reads_info_entry]\n",
    "            line = f.readline()\n",
    "            \n",
    "    return reads_info\n",
    "\n",
    "        \n",
    "def compare_results(fasta_filepath, bwa_filepath, custom_filepath):\n",
    "    with open(fasta_filepath) as f:\n",
    "        f.readline()\n",
    "        fasta_line = f.readline()\n",
    "        \n",
    "    fasta_seq_len = len(fasta_line)\n",
    "    bwa_reads_info = parse_result_file_sam(bwa_filepath)\n",
    "    custom_reads_info = parse_result_file_custom(custom_filepath)\n",
    "    \n",
    "    n_bwa = len(bwa_reads_info)\n",
    "    len_diff_cnt = 0\n",
    "    corr_map_cnt = 0\n",
    "    custom_cnt = 0\n",
    "    \n",
    "    for key in bwa_reads_info:\n",
    "        if key in custom_reads_info:\n",
    "            custom_cnt += 1\n",
    "            if len(bwa_reads_info[key]) != len(custom_reads_info[key]):\n",
    "                len_diff_cnt += 1\n",
    "                \"\"\"print(\"len diff:\")\n",
    "                print(\"bwa: \" + key + \" => \" + str(len(bwa_reads_info[key])))\n",
    "                print(\"cst: \" + key + \" => \" + str(len(custom_reads_info[key])))\"\"\"\n",
    "                for i in range(len(bwa_reads_info[key])):\n",
    "                    for j in range(len(custom_reads_info[key])):\n",
    "                        bwa_entry = bwa_reads_info[key][i]\n",
    "                        custom_entry = custom_reads_info[key][j]\n",
    "                        \n",
    "                        bwa_start = max(0, bwa_entry[\"start_ref\"] - bwa_entry[\"length\"])\n",
    "                        bwa_end = min(fasta_seq_len, bwa_entry[\"start_ref\"] + bwa_entry[\"length\"])\n",
    "                        if bwa_start <= custom_entry[\"start_ref\"] and custom_entry[\"start_ref\"] <= bwa_end:\n",
    "                            corr_map_cnt += 1\n",
    "                        \n",
    "            else:\n",
    "                for i in range(len(bwa_reads_info[key])):\n",
    "                    bwa_entry = bwa_reads_info[key][i]\n",
    "                    custom_entry = custom_reads_info[key][i]\n",
    "                    \n",
    "                    bwa_start = max(0, bwa_entry[\"start_ref\"] - bwa_entry[\"length\"])\n",
    "                    bwa_end = min(fasta_seq_len, bwa_entry[\"start_ref\"] + bwa_entry[\"length\"])\n",
    "                    if bwa_start <= custom_entry[\"start_ref\"] and custom_entry[\"start_ref\"] <= bwa_end:\n",
    "                        corr_map_cnt += 1\n",
    "\n",
    "    print(\"Comparison of position alignment:\\n\")\n",
    "\n",
    "    print(\"Number of entries in bwa: \" + str(n_bwa))\n",
    "    print(\"Number of entries in custom: \" + str(len(custom_reads_info)))\n",
    "    \n",
    "    print(\"% of reads from bwa existing in custom: \" + str(float(custom_cnt) / float(n_bwa)))\n",
    "    print(\"% of reads with correct mapping: \" + str(corr_map_cnt / n_bwa))\n",
    "    print(\"Number of reads with wrong number of appearence: \" + str(len_diff_cnt))\n",
    "\n",
    "    \n",
    "# Program\n",
    "is_test = 1\n",
    "if is_test:\n",
    "    compare_results(\"../data/MT.fa\", \"../data/bwa_res_test.sam\", \"../data/test.fq.samalike\")\n",
    "else:\n",
    "    compare_results(\"../data/MT.fa\", \"../data/bwa_res.sam\", \"../data/art.fq.samalike\")\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
