{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD8CAYAAAB5Pm/hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAD8pJREFUeJzt3X+s3XV9x/HnS+rQ+asULqRpy4qx\ncfjHKNiwGpZFwTlAI/whUWewMV36D1swujjYkrkl+0P/EWdmWBpxlsXJEHU0jKikQswSRVtEBCuj\nss7eldGigNuMP+re++N8Ou/qpffc3nN7zvn0+UhOvt/v53zOOe9777ev8zmf7/d8m6pCktSv5427\nAEnS8jLoJalzBr0kdc6gl6TOGfSS1DmDXpI6Z9BLUucMeknqnEEvSZ1bMe4CAM4666xav379uMuQ\npKmyZ8+ep6pqZqF+ExH069evZ/fu3eMuQ5KmSpJ/G6afUzeS1DmDXpI6Z9BLUucMeknqnEEvSZ0z\n6CWpcwa9JHXOoJekzhn0ktS5ifhm7LRZf8M//b/t/R9445gqkaSFOaKXpM4Z9JLUOYNekjpn0EtS\n54YK+iT7k3wryYNJdre2VUnuSfJYW57R2pPkI0n2JXkoyUXL+QNIko5vMSP611XVxqra1LZvAHZV\n1QZgV9sGuALY0G7bgJtHVawkafGWMnVzFbCjre8Arp7TfmsNfBVYmWT1El5HkrQEwwZ9AV9MsifJ\nttZ2TlU9AdCWZ7f2NcCBOY+dbW2SpDEY9gtTl1TVwSRnA/ck+c5x+maetvqlToM3jG0A55577pBl\nSJIWa6gRfVUdbMtDwOeAi4Enj07JtOWh1n0WWDfn4WuBg/M85/aq2lRVm2ZmFvy/bSVJJ2jBoE/y\noiQvOboOvAF4GNgJbGndtgB3tvWdwDvb2TebgWePTvFIkk6+YaZuzgE+l+Ro/7+vqs8n+Tpwe5Kt\nwPeAa1r/u4ErgX3Aj4B3jbxqSdLQFgz6qnocuGCe9u8Dl83TXsB1I6lOkrRkXr1SOsV49dVTj5dA\nkKTOOaLXsnMEKY2XI3pJ6pxBL0mdM+glqXMGvSR1zoOxkobmgfXp5Ihekjpn0EtS5wx6SeqcQS9J\nnTPoJalzBr0kdc7TK6Ux85RFLTdH9JLUOYNekjrn1I20TJyS0aRwRC9JnTPoJalzBr0kdc45eukE\nOP+uaeKIXpI6Z9BLUuecupFGwKkcTTJH9JLUOYNekjpn0EtS57qco3e+VJJ+ocugV798E5cWz6CX\npDl6HEwMHfRJTgN2A/9eVW9Kch5wG7AKeAC4tqp+muR04Fbg1cD3gbdW1f6RV64TNndH7mEnlibB\nJL9BLGZEfz2wF3hp2/4gcFNV3Zbkb4CtwM1t+XRVvSLJ21q/t46w5mUzyX8oSTpRQ511k2Qt8Ebg\nY207wKXAHa3LDuDqtn5V26bdf1nrL0kag2FH9B8G3ge8pG2fCTxTVUfa9iywpq2vAQ4AVNWRJM+2\n/k/NfcIk24BtAOeee+6J1j/1/BShUXJ/0nwWHNEneRNwqKr2zG2ep2sNcd8vGqq2V9Wmqto0MzMz\nVLGSpMUbZkR/CfDmJFcCL2AwR/9hYGWSFW1UvxY42PrPAuuA2SQrgJcBPxh55ZKkoSwY9FV1I3Aj\nQJLXAn9UVe9I8mngLQzOvNkC3NkesrNtf6Xd/6Wq+qUR/cnmR1pJp6qlXALhj4H3JNnHYA7+ltZ+\nC3Bma38PcMPSSpQkLcWivjBVVfcB97X1x4GL5+nzY+CaEdQmSRoBL2omSZ3zEgjSBPLbyxolR/SS\n1DmDXpI659SNlmQSTludhBqkSeaIXpI654he0kj5CWvyOKKXpM45oj+JHOlIGgdH9JLUOUf0kubl\nJ9B+GPSStIBpf9Mz6Bcw7X9gSTqlg97riYyeb4zS5Dmlg1462Xwj/IVR/i78vR6fQS8twBDRtDPo\nJS3JKKdAnU5dHga9NAX8VKGlMOilKeUct4Zl0GssDBbp5PESCJLUOUf0mpcjbqkfBv2U8GyEk8M3\nOPXIoJfhpqni/rp4Bn1HevwH4CcZaekM+mXSY+hOE3//k8W/x3h51o0kdc6gl6TOOXUjaWI55TMa\njuglqXMLjuiTvAD4MnB6639HVb0/yXnAbcAq4AHg2qr6aZLTgVuBVwPfB95aVfuXqX7f8SVpAcOM\n6H8CXFpVFwAbgcuTbAY+CNxUVRuAp4Gtrf9W4OmqegVwU+snSRqTBUf0VVXAf7XN57dbAZcCv9fa\ndwB/DtwMXNXWAe4A/jpJ2vNoRPwkI2lYQx2MTXIasAd4BfBR4LvAM1V1pHWZBda09TXAAYCqOpLk\nWeBM4KkR1j1xjg1eSZoUQx2MraqfV9VGYC1wMXD+fN3aMse57/8k2ZZkd5Ldhw8fHrZeSdIiLer0\nyqp6Jsl9wGZgZZIVbVS/FjjYus0C64DZJCuAlwE/mOe5tgPbATZt2uS0zjJxiue5+bvRUkzT5TmG\nOetmBvhZC/kXAq9ncID1XuAtDM682QLc2R6ys21/pd3/JefnJU2iU+XNfpgR/WpgR5unfx5we1Xd\nleTbwG1J/hL4BnBL638L8HdJ9jEYyb9tGeqWJA1pmLNuHgIunKf9cQbz9ce2/xi4ZiTVSSM0TR+1\npVHym7GS1DmvdaOJcarMl0onmyN6SeqcQS9JnXPqRlKXjj34Po6pwUmZjnREL0mdM+glqXNO3Who\nk/IxVNLiOKKXpM4Z9JLUOaduJpBTJJJGyRG9JHXOoJekzhn0ktQ55+ilznl5Zjmil6TOGfSS1DmD\nXpI6Z9BLUucMeknqnEEvSZ0z6CWpcwa9JHXOoJekzhn0ktQ5g16SOmfQS1LnDHpJ6pxBL0mdM+gl\nqXMGvSR1bsGgT7Iuyb1J9iZ5JMn1rX1VknuSPNaWZ7T2JPlIkn1JHkpy0XL/EJKk5zbMiP4I8N6q\nOh/YDFyX5FXADcCuqtoA7GrbAFcAG9ptG3DzyKuWJA1twaCvqieq6oG2/p/AXmANcBWwo3XbAVzd\n1q8Cbq2BrwIrk6weeeWSpKEsao4+yXrgQuB+4JyqegIGbwbA2a3bGuDAnIfNtjZJ0hgMHfRJXgx8\nBnh3Vf3weF3naat5nm9bkt1Jdh8+fHjYMiRJizRU0Cd5PoOQ/2RVfbY1P3l0SqYtD7X2WWDdnIev\nBQ4e+5xVtb2qNlXVppmZmROtX5K0gGHOuglwC7C3qj40566dwJa2vgW4c077O9vZN5uBZ49O8UiS\nTr4VQ/S5BLgW+FaSB1vbnwAfAG5PshX4HnBNu+9u4EpgH/Aj4F0jrViStCgLBn1V/TPzz7sDXDZP\n/wKuW2JdkqQR8ZuxktQ5g16SOmfQS1LnDHpJ6pxBL0mdM+glqXMGvSR1zqCXpM4Z9JLUOYNekjpn\n0EtS5wx6SeqcQS9JnTPoJalzBr0kdc6gl6TOGfSS1DmDXpI6Z9BLUucMeknqnEEvSZ0z6CWpcwa9\nJHXOoJekzhn0ktQ5g16SOmfQS1LnDHpJ6pxBL0mdM+glqXMGvSR1bsGgT/LxJIeSPDynbVWSe5I8\n1pZntPYk+UiSfUkeSnLRchYvSVrYMCP6TwCXH9N2A7CrqjYAu9o2wBXAhnbbBtw8mjIlSSdqwaCv\nqi8DPzim+SpgR1vfAVw9p/3WGvgqsDLJ6lEVK0lavBOdoz+nqp4AaMuzW/sa4MCcfrOt7Zck2ZZk\nd5Ldhw8fPsEyJEkLGfXB2MzTVvN1rKrtVbWpqjbNzMyMuAxJ0lEnGvRPHp2SactDrX0WWDen31rg\n4ImXJ0laqhMN+p3Alra+BbhzTvs729k3m4Fnj07xSJLGY8VCHZJ8CngtcFaSWeD9wAeA25NsBb4H\nXNO63w1cCewDfgS8axlqliQtwoJBX1Vvf467LpunbwHXLbUoSdLo+M1YSeqcQS9JnTPoJalzBr0k\ndc6gl6TOGfSS1DmDXpI6Z9BLUucMeknqnEEvSZ0z6CWpcwa9JHXOoJekzhn0ktQ5g16SOmfQS1Ln\nDHpJ6pxBL0mdM+glqXMGvSR1zqCXpM4Z9JLUOYNekjpn0EtS5wx6SeqcQS9JnTPoJalzBr0kdc6g\nl6TOGfSS1LllCfoklyd5NMm+JDcsx2tIkoYz8qBPchrwUeAK4FXA25O8atSvI0kaznKM6C8G9lXV\n41X1U+A24KpleB1J0hCWI+jXAAfmbM+2NknSGKSqRvuEyTXA71bV77fta4GLq+oPj+m3DdjWNl8J\nPLrElz4LeGqJzzFO1j8+01w7WP84jbv2X6uqmYU6rViGF54F1s3ZXgscPLZTVW0Hto/qRZPsrqpN\no3q+k836x2eaawfrH6dpqX05pm6+DmxIcl6SXwHeBuxchteRJA1h5CP6qjqS5A+ALwCnAR+vqkdG\n/TqSpOEsx9QNVXU3cPdyPPdxjGwaaEysf3ymuXaw/nGaitpHfjBWkjRZvASCJHWui6CftksuJPl4\nkkNJHp7TtirJPUkea8szxlnjc0myLsm9SfYmeSTJ9a19Wup/QZKvJflmq/8vWvt5Se5v9f9DO5Fg\nIiU5Lck3ktzVtqep9v1JvpXkwSS7W9tU7DsASVYmuSPJd9q/gddMQ/1TH/RTesmFTwCXH9N2A7Cr\nqjYAu9r2JDoCvLeqzgc2A9e13/e01P8T4NKqugDYCFyeZDPwQeCmVv/TwNYx1riQ64G9c7anqXaA\n11XVxjmnJU7LvgPwV8Dnq+rXgQsY/B0mv/6qmuob8BrgC3O2bwRuHHddQ9S9Hnh4zvajwOq2vhp4\ndNw1Dvlz3An8zjTWD/wq8ADwmwy+9LJivn1qkm4MvpeyC7gUuAvItNTe6tsPnHVM21TsO8BLgX+l\nHducpvqnfkRPP5dcOKeqngBoy7PHXM+CkqwHLgTuZ4rqb1MfDwKHgHuA7wLPVNWR1mWS96EPA+8D\n/qdtn8n01A5QwBeT7Gnfjofp2XdeDhwG/rZNnX0syYuYgvp7CPrM0+apRMssyYuBzwDvrqofjrue\nxaiqn1fVRgaj44uB8+frdnKrWliSNwGHqmrP3OZ5uk5c7XNcUlUXMZhqvS7Jb4+7oEVYAVwE3FxV\nFwL/zSRO08yjh6Af6pILU+DJJKsB2vLQmOt5TkmezyDkP1lVn23NU1P/UVX1DHAfg2MNK5Mc/V7J\npO5DlwBvTrKfwVVhL2Uwwp+G2gGoqoNteQj4HIM32mnZd2aB2aq6v23fwSD4J77+HoK+l0su7AS2\ntPUtDOa+J06SALcAe6vqQ3Pumpb6Z5KsbOsvBF7P4IDavcBbWreJrL+qbqyqtVW1nsF+/qWqegdT\nUDtAkhclecnRdeANwMNMyb5TVf8BHEjyytZ0GfBtpqH+cR8kGNFBkiuBf2Ew1/qn465niHo/BTwB\n/IzBKGErg7nWXcBjbblq3HU+R+2/xWBq4CHgwXa7corq/w3gG63+h4E/a+0vB74G7AM+DZw+7loX\n+DleC9w1TbW3Or/Zbo8c/bc6LftOq3UjsLvtP/8InDEN9fvNWEnqXA9TN5Kk4zDoJalzBr0kdc6g\nl6TOGfSS1DmDXpI6Z9BLUucMeknq3P8CAJYzFNDDsf8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x6d3fa20>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "2000\n",
      "2100\n",
      "2200\n",
      "2300\n",
      "2400\n",
      "done in 2.3367068767547607 s\n"
     ]
    }
   ],
   "source": [
    "from operator import itemgetter\n",
    "import numpy\n",
    "import winsound\n",
    "import time\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "# Turn string MMMMIMMD... into 4M1I2M1D...\n",
    "def compress_letters(word):\n",
    "    res = \"\"\n",
    "    \n",
    "    cur_let = word[0]\n",
    "    cur_cnt = 1\n",
    "    for i in range(1, len(word)):\n",
    "        if word[i] == cur_let:\n",
    "            cur_cnt += 1\n",
    "        else:\n",
    "            res += str(cur_cnt) + cur_let\n",
    "            cur_let = word[i]\n",
    "            cur_cnt = 1\n",
    "            \n",
    "    res += str(cur_cnt) + cur_let\n",
    "    \n",
    "    return res\n",
    "   \n",
    "\n",
    "# Scoring matrix for dynamic programming\n",
    "def scoringMatrix(a,b):\n",
    "    if a==b: return 1\n",
    "    if a=='_' or b=='_': return -7\n",
    "    maxb, minb = max(a,b), min(a,b)\n",
    "    if minb=='A' and maxb=='G': return -1\n",
    "    if minb=='C' and maxb=='T': return -1\n",
    "    return -3\n",
    "\n",
    "\n",
    "# Global alignment of two strings\n",
    "def globalAlignment(x,y, s):\n",
    "    D = numpy.zeros((len(x)+1,len(y)+1), dtype=int)\n",
    "\n",
    "    for i in range(1, len(x)+1):\n",
    "        D[i,0]=D[i-1,0]+s(x[i-1],'_')  \n",
    "    for j in range(1, len(y)+1):\n",
    "        D[0,j]=D[0,j-1]+s('_',y[j-1]) # D[0,j]=D[0,j-1]+s('_',y[i-1])\n",
    "\n",
    "    for i in range(1,len(x)+1):\n",
    "        for j in range(1,len(y)+1):\n",
    "            D[i,j]=max(D[i-1,j]  +s(x[i-1], '_'),\n",
    "                       D[i,j-1]  +s('_',   y[j-1]), \n",
    "                       D[i-1,j-1]+s(x[i-1],y[j-1]))\n",
    "\n",
    "    # function returns table and global alignment score\n",
    "    #alignment score is in cell (n,m) of the matrix\n",
    "    return D, D[len(x),len(y)]\n",
    "    \n",
    "\n",
    "# Traceback through dynamic programming matrix\n",
    "def traceback(x,y,V,s):\n",
    "    # initializing starting position cell(n,m)\n",
    "    i=len(x)\n",
    "    j=len(y)\n",
    "\n",
    "    # initializing strings we use to represent alignments in x, y, edit transcript and global alignment\n",
    "    ax,ay,am, tr = '','','',''\n",
    "\n",
    "    # exit condition is when we reach cell (0,0)\n",
    "    while (i>0 or j>0):\n",
    "        # calculating diagonal, horizontal and vertical scores for current cell\n",
    "        if i>0 and j>0:\n",
    "            sigma = 1 if x[i-1]==y[j-1] else 0\n",
    "            d = V[i-1,j-1] + s(x[i-1],y[j-1])  # diagonal movement   \n",
    "        if i>0 : v = V[i-1,j] + s(x[i-1],'_')  # vertical movement\n",
    "        if j>0 : h = V[i,j-1] + s('_',y[j-1])  # horizontal movement\n",
    "\n",
    "        # backtracing to next(previous) cell\n",
    "        if d>=v and d>=h:\n",
    "            ax+=x[i-1]\n",
    "            ay+=y[j-1]\n",
    "            if sigma==1:\n",
    "                tr+='M'\n",
    "                am+='|'\n",
    "            else:\n",
    "                tr+='R'\n",
    "                am+=' '\n",
    "            i-=1\n",
    "            j-=1\n",
    "        elif v>=h:\n",
    "            ax+=x[i-1]\n",
    "            ay+='_'\n",
    "            tr+='I'\n",
    "            am+=' '\n",
    "            i-=1\n",
    "        else:\n",
    "            ay+=y[j-1]\n",
    "            ax+='_'\n",
    "            tr+='D'\n",
    "            am+=' '\n",
    "            j-=1\n",
    "\n",
    "    alignment='\\n'.join([ax[::-1][0:50], am[::-1][0:50], ay[::-1][0:50]])\n",
    "    return alignment, tr[::-1]\n",
    "\n",
    "\n",
    "# Index structure, hash map\n",
    "class IndexHash(object):\n",
    "    \n",
    "    def __init__(self, t, ln):\n",
    "        \"\"\" Create index, extracting substrings of length 'ln' \"\"\"\n",
    "        self.t = t\n",
    "        self.ln = ln\n",
    "        self.index = {}\n",
    "        for i in range(len(t) - ln + 1):\n",
    "            substr = t[i: i + ln]\n",
    "            if substr in self.index:\n",
    "                self.index[substr].append(i) # substring already in dictionary\n",
    "            else:\n",
    "                self.index[substr] = [i] # add to dictionary\n",
    "    \n",
    "    def query(self, p):\n",
    "        \"\"\" Return candidate alignments for p \"\"\"\n",
    "        return self.index.get(p[:self.ln], [])\n",
    "\n",
    "    \n",
    "# Make a complement from the argument string\n",
    "def make_strand_complement(strand):\n",
    "    complements = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "    return ''.join([complements.get(base, base) for base in strand.upper()[::-1]])      \n",
    "\n",
    "\n",
    "# Class for alignment\n",
    "class CustomAligner(object):\n",
    "    \n",
    "    def __init__(self, fasta_fp, fastq_fp, seed_ln, seed_int):\n",
    "        self.fasta_filepath = fasta_fp\n",
    "        self.fastq_filepath = fastq_fp\n",
    "        self.seed_length = seed_ln\n",
    "        self.seed_interval = seed_int\n",
    "        self.seed_win_size = 300 # default\n",
    "        self.fasta_name = \"\"\n",
    "        self.fasta_sequence = \"\"\n",
    "        self.read_entries = dict()\n",
    "        self.seed_windows = dict()\n",
    "\n",
    "    # Read fasta file and save sequence value \n",
    "    def parse_fasta(self):\n",
    "        with open(self.fasta_filepath) as f:\n",
    "            self.fasta_name = f.readline().rstrip()[1:]\n",
    "            self.fasta_sequence = f.readline().rstrip()\n",
    "        \n",
    "        for i in range(100, 300):\n",
    "            if len(self.fasta_sequence) % i == 0:\n",
    "                self.seed_win_size = i\n",
    "                break\n",
    "        \n",
    "    # Read fastq file and save reads in a dict: read_name -> (read, read_complement)\n",
    "    def parse_fastq(self):\n",
    "        with open(self.fastq_filepath) as f:\n",
    "            line = f.readline().rstrip()\n",
    "            while line:\n",
    "                if line.startswith(\"@\") or line.startswith(\"+\"):\n",
    "                    prev = line[0]\n",
    "                    if line.startswith(\"@\"):\n",
    "                        name = line[1:].split(\"/\")[0]\n",
    "                else:\n",
    "                    if prev.startswith(\"@\"):\n",
    "                        strand = line\n",
    "                    elif prev.startswith(\"+\"):\n",
    "                        self.read_entries[name] = dict()\n",
    "                        self.read_entries[name][\"strand\"] = strand\n",
    "                        self.read_entries[name][\"reverse_strand\"] = make_strand_complement(strand)\n",
    "                        self.read_entries[name][\"strand_quality\"] = line\n",
    "                        self.read_entries[name][\"blocks\"] = list()\n",
    "                        self.read_entries[name][\"fasta_name\"] = self.fasta_name\n",
    "                        self.read_entries[name][\"positions_in_ref\"] = list()\n",
    "                        self.read_entries[name][\"cigars\"] = list()\n",
    "                        self.read_entries[name][\"flags\"] = 4 # Default unmapped\n",
    "\n",
    "                line = f.readline().rstrip()\n",
    "          \n",
    "    # Make seeds from given read and save info\n",
    "    def align_seeds_from_read(self, read_string, index):\n",
    "        seed_info = list()\n",
    "        for i in range (0, len(read_string), self.seed_interval):\n",
    "            end = min(len(read_string), i + self.seed_length)\n",
    "            cur_seed = read_string[i: end]\n",
    "            positions_list = index.query(cur_seed)\n",
    "            if len(positions_list) > 0:\n",
    "                seed_info.append((positions_list, i, end))\n",
    "        \n",
    "        return seed_info\n",
    "    \n",
    "    # Split each read into seeds, and save info of it \n",
    "    def make_seeds(self, index_forward, index_reverse):\n",
    "        # (list_of_positions_in_ref, start_pos_in_read, end_pos_in_read)\n",
    "        for key, value in self.read_entries.items():\n",
    "            tmp_lists = list()\n",
    "            tmp_lists.append(self.align_seeds_from_read(value[\"strand\"], index_forward))\n",
    "            #tmp_lists.append(self.align_seeds_from_read(value[0], index_reverse))\n",
    "            tmp_lists.append(self.align_seeds_from_read(value[\"reverse_strand\"], index_forward))\n",
    "            #tmp_lists.append(self.align_seeds_from_read(value[1], index_reverse))\n",
    "            max_len, max_list, max_ind = max((len(x), x, tmp_lists.index(x)) for x in tmp_lists)\n",
    "            \n",
    "            if max_len != 0:\n",
    "                self.read_entries[key][\"flags\"] = max_ind\n",
    "            \n",
    "            for seed_entry in max_list:\n",
    "                for position in seed_entry[0]:\n",
    "                    block_entry = dict()\n",
    "                    block_entry[\"pos_ref\"] = int(position)\n",
    "                    block_entry[\"start_read\"] = seed_entry[1]\n",
    "                    block_entry[\"end_read\"] = seed_entry[2]\n",
    "                    \n",
    "                    win_key = int(position / self.seed_win_size)\n",
    "                    if win_key in self.seed_windows:\n",
    "                        self.seed_windows[win_key] += 1\n",
    "                    else:\n",
    "                        self.seed_windows[win_key] = 1\n",
    "                    self.read_entries[key][\"blocks\"].append(block_entry)\n",
    "       \n",
    "    # Merge seeds where can be merged\n",
    "    def merge_seeds(self):\n",
    "        # Merge seeds within a read\n",
    "        for key in self.read_entries:\n",
    "            self.read_entries[key][\"blocks\"] = sorted(self.read_entries[key][\"blocks\"], key = itemgetter(\"pos_ref\"))\n",
    "            if len(self.read_entries[key][\"blocks\"]) > 0:\n",
    "                prev_block = self.read_entries[key][\"blocks\"][0]\n",
    "                i = 1\n",
    "                while i < len(self.read_entries[key][\"blocks\"]):\n",
    "                    if prev_block[\"end_read\"] >= self.read_entries[key][\"blocks\"][i][\"start_read\"] and \\\n",
    "                    prev_block[\"pos_ref\"] + prev_block[\"end_read\"] - prev_block[\"start_read\"] >= \\\n",
    "                    self.read_entries[key][\"blocks\"][i][\"pos_ref\"] and \\\n",
    "                    prev_block[\"pos_ref\"] + prev_block[\"end_read\"] - prev_block[\"start_read\"] <= len(self.fasta_sequence):\n",
    "                        self.read_entries[key][\"blocks\"][i - 1][\"end_read\"] = max(self.read_entries[key][\"blocks\"][i - 1][\"end_read\"],\n",
    "                                                                                  self.read_entries[key][\"blocks\"][i][\"end_read\"])\n",
    "                        self.read_entries[key][\"blocks\"].remove(self.read_entries[key][\"blocks\"][i])\n",
    "                        prev_block = self.read_entries[key][\"blocks\"][i - 1]\n",
    "                    else:\n",
    "                        prev_block = self.read_entries[key][\"blocks\"][i]\n",
    "                        i += 1\n",
    "           \n",
    "    # Perform seed\n",
    "    def seed(self):\n",
    "        index_forward = IndexHash(self.fasta_sequence, self.seed_length)\n",
    "        index_reverse = IndexHash(make_strand_complement(self.fasta_sequence), self.seed_length)\n",
    "        \n",
    "        self.make_seeds(index_forward, index_reverse)\n",
    "        \n",
    "        for i in range(int(len(self.fasta_sequence) / self.seed_win_size) + 1):\n",
    "            if i not in self.seed_windows:\n",
    "                self.seed_windows[i] = 0\n",
    "                \n",
    "        plt.bar(list(self.seed_windows.keys()), list(self.seed_windows.values()))\n",
    "        plt.show()\n",
    "            \n",
    "        self.merge_seeds()\n",
    "    \n",
    "    # Do dynamic programming over defined read\n",
    "    def dyn_prog(self, read_key, read_start, pos_in_ref):\n",
    "        value = self.read_entries[read_key]\n",
    "        \n",
    "        soft_clip_start = \"\"\n",
    "        soft_clip_end = \"\"\n",
    "\n",
    "        y = value[\"strand\"] if value[\"flags\"] == 0 else value[\"reverse_strand\"]\n",
    "\n",
    "        start_ref = pos_in_ref - read_start\n",
    "        if start_ref < 0:\n",
    "            start_ref = 0\n",
    "            soft_clip_start = str(abs(pos_in_ref - read_start)) + \"S\"\n",
    "\n",
    "        end_ref = start_ref + len(value[\"strand\"])\n",
    "        if end_ref > len(self.fasta_sequence):\n",
    "            end_ref = len(self.fasta_sequence)\n",
    "            soft_clip_end = str(start_ref + len(value[\"strand\"]) - len(self.fasta_sequence)) + \"S\"\n",
    "            y = y[:end_ref - start_ref]\n",
    "\n",
    "        x = self.fasta_sequence[start_ref: end_ref]\n",
    "\n",
    "        D, alignment_score = globalAlignment(x, y, scoringMatrix)\n",
    "        alignment, transcript = traceback(x,y, D, scoringMatrix)\n",
    "        \n",
    "        return alignment_score, soft_clip_start + compress_letters(transcript) + soft_clip_end, start_ref\n",
    "    \n",
    "    # Adding info about reads\n",
    "    def add_info(self, read_key, start_ref, alignment_score, transcript):\n",
    "        self.read_entries[read_key][\"positions_in_ref\"].append(start_ref)\n",
    "        self.read_entries[read_key][\"map_qual\"] = alignment_score\n",
    "        self.read_entries[read_key][\"cigars\"].append(transcript)\n",
    "    \n",
    "    # Wrapping functions\n",
    "    def dyn_pr_add_info(self, read_key, read_start, pos_in_ref):\n",
    "        alignment_score, transcript, start_ref = self.dyn_prog(read_key, read_start, pos_in_ref)\n",
    "        self.add_info(read_key, start_ref, alignment_score, transcript)\n",
    "    \n",
    "    def find_best_candidate(self, read_key):\n",
    "        blocks = self.read_entries[read_key][\"blocks\"]\n",
    "        blocks_ranking = dict()\n",
    "        for i in range(len(blocks)):\n",
    "            block = blocks[i]\n",
    "            pos = block[\"pos_ref\"] + int((block[\"end_read\"] - block[\"start_read\"]) / 2)\n",
    "            blocks_ranking[i] = self.seed_windows[int(pos / self.seed_win_size)]\n",
    "            \n",
    "        blocks_ranking = sorted(blocks_ranking.items(), key=itemgetter(1))[::-1]\n",
    "        #print(blocks_ranking)\n",
    "        candidates = blocks_ranking[0: min(3, len(blocks_ranking))]\n",
    "        best_cand = 0\n",
    "        best_tuple = (-100000, \"\", 0)\n",
    "        for cand in candidates:\n",
    "            (alignment_score, transcript, start_ref) = self.dyn_prog(read_key, blocks[cand[0]][\"start_read\"], \\\n",
    "                                                                    blocks[cand[0]][\"pos_ref\"])\n",
    "            if alignment_score > best_tuple[0]:\n",
    "                best_tuple = (alignment_score, transcript, start_ref)\n",
    "                best_cand = cand\n",
    "        \n",
    "        self.add_info(read_key, best_tuple[2], best_tuple[0], best_tuple[1])\n",
    "    \n",
    "    def extend(self):\n",
    "        c = 0\n",
    "        for key, value in self.read_entries.items():\n",
    "            c += 1\n",
    "            if c % 100 == 0:\n",
    "                print(c)\n",
    "                \n",
    "            blocks = value[\"blocks\"]\n",
    "            n_bl = len(blocks)\n",
    "            strand_len = len(value[\"strand\"])\n",
    "            \n",
    "            # Unmapped\n",
    "            if n_bl == 0:\n",
    "                self.add_info(key, 0, 0, \"*\")\n",
    "            else:\n",
    "                # One block in read, save if completely aligned, do dynamic programming if not\n",
    "                if n_bl == 1:\n",
    "                    if blocks[0][\"start_read\"] == 0 and blocks[0][\"end_read\"] == len(value[\"strand\"]):\n",
    "                        self.add_info(key, blocks[0][\"pos_ref\"], strand_len, str(strand_len) + \"M\")\n",
    "                    else:\n",
    "                        self.dyn_pr_add_info(key, blocks[0][\"start_read\"], blocks[0][\"pos_ref\"])\n",
    "                else:\n",
    "                    # Two or more blocks in read, if diff between first and last block is acceptable, do dyn prog\n",
    "                    if (blocks[n_bl - 1][\"pos_ref\"] + blocks[n_bl - 1][\"end_read\"] - blocks[n_bl - 1][\"start_read\"]) - \\\n",
    "                    (blocks[0][\"pos_ref\"] - blocks[0][\"start_read\"]) <= len(value[\"strand\"]) + 10:\n",
    "                        self.dyn_pr_add_info(key, blocks[0][\"start_read\"], blocks[n_bl - 1][\"pos_ref\"])\n",
    "                    else:\n",
    "                        # If there are edging blocks \n",
    "                        if blocks[0][\"pos_ref\"] <= len(value[\"strand\"]) and \\\n",
    "                        blocks[n_bl - 1][\"pos_ref\"] >= len(self.fasta_sequence) - len(value[\"strand\"]):\n",
    "                            self.dyn_pr_add_info(key, blocks[0][\"start_read\"], blocks[0][\"pos_ref\"])\n",
    "                            self.dyn_pr_add_info(key, blocks[n_bl - 1][\"start_read\"], blocks[n_bl - 1][\"pos_ref\"])\n",
    "                        else:\n",
    "                            self.find_best_candidate(key)\n",
    "        \n",
    "    def save_to_file(self, filepath):\n",
    "        with open(filepath, 'w') as file:\n",
    "            for key, value in self.read_entries.items():\n",
    "                for i in range (len(value[\"positions_in_ref\"])):\n",
    "                #for pos in value[\"positions_in_ref\"]:\n",
    "                    file.write(key + \"\\t\" + str(value[\"flags\"]) + \"\\t\" + str(value[\"fasta_name\"]) + \"\\t\" +\n",
    "                               str(value[\"positions_in_ref\"][i]) + \"\\t\" + str(value[\"map_qual\"]) + \"\\t\" +\n",
    "                               str(value[\"cigars\"][i]) + \"\\t\" + str(value[\"strand\"]) + \"\\t\" +\n",
    "                               str(value[\"reverse_strand\"]) + \"\\n\")\n",
    "\n",
    "# Program\n",
    "start_time = time.time()\n",
    "#fastq_filepath = \"data/test.fq\"\n",
    "fastq_filepath = \"../data/art.fq\"\n",
    "ca = CustomAligner(\"../data/MT.fa\", fastq_filepath, 20, 10)\n",
    "ca.parse_fasta()\n",
    "ca.parse_fastq()\n",
    "ca.seed()\n",
    "ca.extend()\n",
    "ca.save_to_file(fastq_filepath + \".samalike\")\n",
    "\n",
    "print(\"done in \" + str(time.time() - start_time) + \" s\")\n",
    "winsound.PlaySound(\"*\", winsound.SND_ALIAS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
